openwisp_controller/config/static/config/js/widget.js

```diff
@@ -153,6 +153,49 @@
         return error;
     };
 
+    var getSystemDefinedVariables = function () {
+        var regexp = new RegExp("(.*?):.(.*)");
+        var systemDefinedVariables = {};
+        $("#system-context").children().map((index, element) => {
+            var text = $(element).text();
+            if (regexp.test(text)) {
+                var matched = text.match(regexp);
+                systemDefinedVariables[matched[1]] = matched[2];
+            }
+        });
+        return systemDefinedVariables;
+    };
+    
+    var validateMaxLength = function () {
+        var systemDefinedVariables = getSystemDefinedVariables();
+        var configVariables = {};
+        if (window.getContext()) {
+            configVariables = JSON.parse(window.getContext().value);
+        }
+        var allVariables = Object.assign({}, systemDefinedVariables, configVariables);
+        var variablesKeys = Object.keys(allVariables);
+        var allFields = $(".jsoneditor .vTextField");
+        allFields.map((index, field) => {
+            if ($(field).attr('maxLength')) {
+                var maxLength = $(field).attr('maxLength');
+                $(field).on('input', () => {
```

Comment:
The device I was testing this has total of 13 system defined variables and context variables. When I entered just `{` in the input field, the regex got evaluated 13 times and the maxLength attribute was set. Will it possible to use [debouncing](https://www.geeksforgeeks.org/debouncing-in-javascript/) here? 🤔 

@nemesisdesign we have that compounding effect here as well. After changing configuration elements for some time, the number of times this  getting executed increased. When I added a new interface, the change handler also got added to fields to which it was already added.

Reply:
Thanks @pandafy for spotting this, yes totally, I think `updateRaw()` is called often, so we have to make sure we do not bind the same handler multiple times, can you double check this @codesankalp please?

Reply:
@nemesisdesign and @pandafy I have made the changes. Needs your review.